master/addexptrigger.sql

```diff
@@ -0,0 +1,150 @@
+        PERFORM stream_experiment_notify(NULL, NULL, to_jsonb(NEW), work);
+    ELSEIF (TG_OP = 'UPDATE') THEN
+        proj = project_id from experiments where experiments.id = NEW.experiment_id;
+work = workspace_id from projects where projects.id = proj;
+        PERFORM stream_experiment_notify(to_jsonb(OLD), work, to_jsonb(NEW), work);
+    ELSEIF (TG_OP = 'DELETE') THEN
+        proj = project_id from experiments where experiments.id = OLD.experiment_id;
+work = workspace_id from projects where projects.id = proj;
+        PERFORM stream_experiment_notify(to_jsonb(OLD), work, NULL, NULL);
+        -- DELETEs trigger BEFORE, and must return a non-NULL value.
+return OLD;
+END IF;
+return NULL;
+END;
+$$ LANGUAGE plpgsql;
+-- INSERT and UPDATE should fire AFTER to guarantee to emit the final row value.
+DROP TRIGGER IF EXISTS stream_experiment_trigger_iu ON experiments;
+CREATE TRIGGER stream_experiment_trigger_iu
+    AFTER INSERT OR UPDATE OF
+    state, archived, progress, start_time, end_time, notes
+                    ON experiments
+                        FOR EACH ROW EXECUTE PROCEDURE stream_experiment_change();
+-- DELETE should fire BEFORE to guarantee the experiment still exists to grab the workspace_id.
+DROP TRIGGER IF EXISTS stream_experiment_trigger_d ON experiments;
+CREATE TRIGGER stream_experiment_trigger_d
+    BEFORE DELETE ON experiments
+    FOR EACH ROW EXECUTE PROCEDURE stream_experiment_change();
+
+-- Trigger for detecting experiment permission scope changes derived from projects.workspace_id.
+CREATE OR REPLACE FUNCTION stream_experiment_workspace_change_notify() RETURNS TRIGGER AS $$
+DECLARE
+exp RECORD;
+    jexp jsonb;
+BEGIN
+FOR experiment IN
+SELECT
+    e.*,
+    workspace_id
+FROM
+    projects p
+        INNER JOIN
+    experiments e ON e.project_id = p.id
+WHERE
+        p.id = NEW.id
+    LOOP
+        experiment.seq = nextval('stream_experiment_seq');
+UPDATE experiments SET seq = exp.seq where id = exp.id;
+jexp = to_jsonb(exp);
+        PERFORM stream_experiment_notify(jexp, OLD.workspace_id, jexp, NEW.workspace_id);
+END LOOP;
+    -- return value for AFTER triggers is ignored
+return NULL;
+END;
+$$ LANGUAGE plpgsql;
+--
+DROP TRIGGER IF EXISTS stream_experiment_workspace_change_trigger ON projects;
+CREATE TRIGGER stream_experiment_workspace_change_trigger
+    AFTER UPDATE OF workspace_id ON projects
+    FOR EACH ROW EXECUTE PROCEDURE stream_experiment_workspace_change_notify();
+
+-- Trigger for detecting experiment permission scope changes derived from experiment.project_id.
+CREATE OR REPLACE FUNCTION stream_experiment_project_change_notify() RETURNS TRIGGER AS $$
+DECLARE
+exp RECORD;
+    jexp jsonb;
+    proj integer;
+    oldwork integer;
+    newwork integer;
+BEGIN
+FOR exp IN
+SELECT
+    e.*
+FROM
+    experiments e
+WHERE
+        e.id = NEW.id
+    LOOP
+        proj = project_id from experiments where experiments.id = OLD.id;
```

Comment:
Same as above these can be simplified to one line
```suggestion
        work = workspace_id from projects where projects.id = OLD.project_id;
```

Reply:
Actually ignore this, I don't think we actually need this trigger since we're already tracking changes on the experiment table with our other triggers

Reply:
would we have to modify the stream_trial version of this to also update the experiment seq and call `stream_experiment_notify`?

Reply:
Nope, because the intention is to make to force an update if this causes the trial to fall out of the user's permission scope

Reply:
right, makes sense! gotta love the bootem strat lol